File: apps/api/src/auth/auth.service.ts
import {
  Injectable,
  UnauthorizedException,
  BadRequestException,
  Logger,
} from '@nestjs/common';
import * as bcrypt from 'bcryptjs';
import * as crypto from 'crypto';
import { PrismaService } from '../common/prisma.service';
import { RedisService } from '../redis.service';
import { AuditService } from '../audit/audit.service';
import { LoginDto, ChangePasswordDto } from './auth.dto';
import { ErrorCode } from '@shaliach/shared';

@Injectable()
export class AuthService {
  private readonly logger = new Logger(AuthService.name);
  private readonly SESSION_TTL_SECONDS = 7 * 24 * 60 * 60; // 7 days

  constructor(
    private prisma: PrismaService,
    private redis: RedisService,
    private audit: AuditService,
  ) {}

  async login(dto: LoginDto, ipAddress?: string, userAgent?: string) {
    const user = await this.prisma.user.findUnique({
      where: { email: dto.email.toLowerCase().trim() },
    });

    if (!user || !user.isActive) {
      this.logger.warn(`Failed login attempt for email: ${dto.email} from IP: ${ipAddress}`);
      throw new UnauthorizedException({
        code: ErrorCode.AUTH_INVALID_CREDENTIALS,
        message: 'Invalid email address or password',
      });
    }

    const isPasswordValid = await bcrypt.compare(dto.password, user.passwordHash);
    if (!isPasswordValid) {
      this.logger.warn(`Failed login password for user: ${user.email} from IP: ${ipAddress}`);
      throw new UnauthorizedException({
        code: ErrorCode.AUTH_INVALID_CREDENTIALS,
        message: 'Invalid email address or password',
      });
    }

    // Generate secure random session ID
    const sessionId = crypto.randomBytes(32).toString('hex');
    await this.redis.set(`session:${sessionId}`, user.id, this.SESSION_TTL_SECONDS);

    await this.audit.log({
      userId: user.id,
      action: 'USER_LOGIN',
      entityType: 'User',
      entityId: user.id,
      ipAddress,
      userAgent,
    });

    this.logger.log(`Successful login for user: ${user.email}`);

    return {
      sessionId,
      user: {
        id: user.id,
        email: user.email,
        name: user.name,
      },
    };
  }

  async logout(sessionId: string, userId?: string, ipAddress?: string) {
    if (sessionId) {
      await this.redis.del(`session:${sessionId}`);
    }

    if (userId) {
      await this.audit.log({
        userId,
        action: 'USER_LOGOUT',
        entityType: 'User',
        entityId: userId,
        ipAddress,
      });
    }

    return { success: true, message: 'Logged out successfully' };
  }

  async changePassword(userId: string, dto: ChangePasswordDto, ipAddress?: string) {
    const user = await this.prisma.user.findUnique({
      where: { id: userId },
    });

    if (!user) {
      throw new UnauthorizedException({
        code: ErrorCode.UNAUTHORIZED,
        message: 'User not found',
      });
    }

    const isCurrentValid = await bcrypt.compare(dto.currentPassword, user.passwordHash);
    if (!isCurrentValid) {
      throw new BadRequestException({
        code: ErrorCode.VALIDATION_FAILED,
        message: 'Current password does not match',
      });
    }

    const newHash = await bcrypt.hash(dto.newPassword, 12);
    await this.prisma.user.update({
      where: { id: userId },
      data: { passwordHash: newHash },
    });

    await this.audit.log({
      userId,
      action: 'PASSWORD_CHANGED',
      entityType: 'User',
      entityId: userId,
      ipAddress,
    });

    return { success: true, message: 'Password changed successfully' };
  }
}

File: apps/api/app/services/auth_service.py
"""Authentication service."""

import logging
import secrets
import bcrypt
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from redis.asyncio import Redis

from ..models.user import User
from ..schemas.auth import LoginSchema, ChangePasswordSchema
from ..errors import AuthenticationError, ValidationError, ErrorCode
from .audit_service import AuditService

logger = logging.getLogger("shaliach.auth")


class AuthService:
    SESSION_TTL_SECONDS = 7 * 24 * 60 * 60  # 7 days

    def __init__(self, db: AsyncSession, redis: Redis, audit: AuditService):
        self.db = db
        self.redis = redis
        self.audit = audit

    async def login(
        self,
        dto: LoginSchema,
        ip_address: str | None = None,
        user_agent: str | None = None,
    ) -> dict:
        email = str(dto.email).lower().strip()
        stmt = select(User).where(User.email == email)
        result = await self.db.execute(stmt)
        user = result.scalar_one_or_none()

        if not user or not user.is_active:
            logger.warning(f"Failed login attempt for email: {dto.email} from IP: {ip_address}")
            raise AuthenticationError(
                code=ErrorCode.AUTH_INVALID_CREDENTIALS,
                message="Invalid email address or password",
            )

        # Verify password
        try:
            password_matches = bcrypt.checkpw(
                dto.password.encode("utf-8"),
                user.password_hash.encode("utf-8") if isinstance(user.password_hash, str) else user.password_hash,
            )
        except Exception:
            password_matches = False

        if not password_matches:
            logger.warning(f"Failed login password for user: {user.email} from IP: {ip_address}")
            raise AuthenticationError(
                code=ErrorCode.AUTH_INVALID_CREDENTIALS,
                message="Invalid email address or password",
            )

        # Generate secure random session ID
        session_id = secrets.token_hex(32)
        await self.redis.set(f"session:{session_id}", user.id, ex=self.SESSION_TTL_SECONDS)

        await self.audit.log(
            action="USER_LOGIN",
            entity_type="User",
            entity_id=user.id,
            user_id=user.id,
            ip_address=ip_address,
            user_agent=user_agent,
        )

        logger.info(f"Successful login for user: {user.email}")

        return {
            "sessionId": session_id,
            "user": {
                "id": user.id,
                "email": user.email,
                "name": user.name,
            },
        }

    async def logout(
        self,
        session_id: str | None,
        user_id: str | None = None,
        ip_address: str | None = None,
    ) -> dict:
        if session_id:
            await self.redis.delete(f"session:{session_id}")

        if user_id:
            await self.audit.log(
                action="USER_LOGOUT",
                entity_type="User",
                entity_id=user_id,
                user_id=user_id,
                ip_address=ip_address,
            )

        return {"success": True, "message": "Logged out successfully"}

    async def change_password(
        self,
        user_id: str,
        dto: ChangePasswordSchema,
        ip_address: str | None = None,
    ) -> dict:
        stmt = select(User).where(User.id == user_id)
        result = await self.db.execute(stmt)
        user = result.scalar_one_or_none()

        if not user:
            raise AuthenticationError(
                code=ErrorCode.UNAUTHORIZED,
                message="User not found",
            )

        try:
            is_current_valid = bcrypt.checkpw(
                dto.currentPassword.encode("utf-8"),
                user.password_hash.encode("utf-8") if isinstance(user.password_hash, str) else user.password_hash,
            )
        except Exception:
            is_current_valid = False

        if not is_current_valid:
            raise ValidationError(
                message="Current password does not match",
            )

        new_hash = bcrypt.hashpw(dto.newPassword.encode("utf-8"), bcrypt.gensalt(12)).decode("utf-8")
        user.password_hash = new_hash
        await self.db.commit()

        await self.audit.log(
            action="PASSWORD_CHANGED",
            entity_type="User",
            entity_id=user_id,
            user_id=user_id,
            ip_address=ip_address,
        )

        return {"success": True, "message": "Password changed successfully"}
